### PR TITLE
connect: strip port from DNS SANs for ingress gateway leaf cert

### PR DESCRIPTION
**Pull request by kyhavlov**
Closes https://github.com/hashicorp/consul/issues/11092 - currently we're passing along whatever's in the `Hosts` field verbatim to construct the list of SANs. This PR changes that to strip the `:` and everything after it from each `Host` string so that it's in the format Vault expects when signing the leaf cert.

### CODE REVIEW

File: agent/proxycfg/ingress_gateway.go [hunk 1]
```diff
@@ -288,7 +289,19 @@ func (s *handlerIngressGateway) generateIngressDNSSANs(snap *ConfigSnapshot) []s
...
 
-	dnsNames = append(dnsNames, snap.IngressGateway.Hosts...)
+	addedHosts := make(map[string]struct{})
+	for _, host := range snap.IngressGateway.Hosts {
+		if _, ok := addedHosts[host]; ok {
+			continue
+		}
+
+		// Drop everything after the ':' from the host when constructing the DNS SANs.
+		hostSegments := strings.Split(host, ":")
```
[the remark this answers is not in the document]
Comment (kyhavlov) [reply agent/proxycfg/ingress_gateway.go:299]: The list of hosts also ends up in the `domains` field of the `VirtualHosts` we send to envoy, which needs to include the port in some cases: https://www.envoyproxy.io/docs/envoy/latest/faq/debugging/why_is_my_route_not_found 

So including the port is needed for the route, but causes an error when signing the cert if it's included in the DNS SANs

**Review by hashi-derek — APPROVED** [review]

**Comment by jkirschner-hashicorp** [comment]
@kyhavlov : Thanks! A few follow-up questions:
- Which versions do we intend to attempt to backport this to?
- Should we intervene where you did (at the ingress gateway point)? Or is there any reason to strip host out from dnsnames at a more general point ([like in the CSR code](https://github.com/hashicorp/consul/blob/325c1c0dd7af97601edd72601b4ab4e190b25b00/agent/connect/csr.go#L54))?
- Do we need any automated tests for this? (Since there are no unit tests, I assume we manually tested instead?)

File: agent/connect/csr.go [hunk 2]
```diff
@@ -47,11 +48,28 @@ func SigAlgoForKeyType(keyType string) x509.SignatureAlgorithm {
 // along with the PEM-encoded private key for this certificate.
 func CreateCSR(uri CertURI, privateKey crypto.Signer,
 	dnsNames []string, ipAddresses []net.IP, extensions ...pkix.Extension) (string, error) {
+
+	// Drop everything after the ':' from the name when constructing the DNS SANs.
+	uniqueNames := make(map[string]struct{})
+	formattedDNSNames := make([]string, 0)
+	for _, host := range dnsNames {
+		hostSegments := strings.Split(host, ":")
+		formattedHost := hostSegments[0]
```
Comment (jkirschner-hashicorp) [inline agent/connect/csr.go:57]: Should we confirm that `len(hostSegments) > 0` here before accessing the 0th element, to ensure we don't have an out-of-bounds access?

File: agent/connect/csr.go [hunk 3]
```diff
@@ -47,11 +48,28 @@ func SigAlgoForKeyType(keyType string) x509.SignatureAlgorithm {
...
+		if _, ok := uniqueNames[formattedHost]; ok {
+			continue
+		}
+
+		if len(hostSegments) >= 1 {
```
Comment (jkirschner-hashicorp) [inline agent/connect/csr.go:62]: This check seems to be ineffective in its current position, because we're already assuming that `hostSegments` has at least 1 element by accessing its 0th element without a guard.

**Review by jkirschner-hashicorp — CHANGES_REQUESTED** [review]
@kyhavlov: I _think_ I spotted something that can quickly be fixed.

File: agent/connect/csr.go (hunk 2 above)
Comment (jkirschner-hashicorp) [reply agent/connect/csr.go:57]: To test this edge case, I guess you could add a host `""` to the array in your test

Comment (pglass) [reply agent/connect/csr.go:57]: This should be okay. [`strings.Split`](https://pkg.go.dev/strings#Split) only returns an empty list when both arguments are empty.

`formattedHost` could be empty here, but that would only occur if the one of `dnsNames` had no hostname, like `:8080`.

File: agent/connect/csr_test.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	require.NoError(t, err)
+	spiffeID := &SpiffeIDService{
+		Host:       "7528f42f-92e5-4db4-b84c-3405c3ca91e6",
+		Service:    "srv1",
+		Datacenter: "dc1",
+	}
+	csr, err := CreateCSR(spiffeID, pk, []string{
+		"foo.example.com",
+		"foo.example.com:8080",
+		"bar.example.com",
```
Comment (pglass) [inline agent/connect/csr_test.go:20]: Maybe throw in a wildcard hostname in this test, like `*.example.com` because [that is allowed](https://developer.hashicorp.com/consul/docs/connect/config-entries/ingress-gateway#hosts)

**Review by pglass — APPROVED** [review]

File: agent/connect/csr.go (hunk 2 above)
Comment (jkirschner-hashicorp) [reply agent/connect/csr.go:57]: We wouldn't want a bad configuration (like "`:8080`") to cause a panic though, right?

**Review by jkirschner-hashicorp — APPROVED** [review]
LGTM!
